core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SimpleEditorManagerService.js

```diff
@@ -207,37 +241,72 @@ oppia.factory('SimpleEditorManagerService', [
           return;
         }
         var nextStateName = state.interaction.answerGroups[0].outcome.dest;
-        var allStateNames = SimpleEditorShimService.getAllStateNames();
         // Change init state name, if init_state is being deleted.
         if (SimpleEditorShimService.getInitStateName() === stateName) {
           explorationInitStateNameService.displayed = nextStateName;
           explorationInitStateNameService.saveDisplayedValue(nextStateName);
         }
+        redirectAllIncomingNodes(stateName, nextStateName);
+        SimpleEditorShimService.saveStateContent(nextStateName, state.content);
+        SimpleEditorShimService.deleteState(stateName);
+        data.questionList.removeQuestion(question);
+      },
 
-        for (var i = 0; i < allStateNames.length; i++) {
-          var currentState = SimpleEditorShimService
-            .getState(allStateNames[i]);
-          var newAnswerGroups = currentState.interaction.answerGroups;
-          var answerGroupsHaveChanged = false;
-          currentState.interaction.answerGroups.forEach(function(answerGroup,
-            idx) {
-            if (answerGroup.outcome.dest === stateName) {
-              newAnswerGroups[idx].outcome.dest = nextStateName;
-              answerGroupsHaveChanged = true;
-            }
-          });
-          if (answerGroupsHaveChanged) {
-            SimpleEditorShimService
-              .saveAnswerGroups(allStateNames[i], newAnswerGroups);
-            data.questionList.getBindableQuestion(allStateNames[i])
-              .setAnswerGroups(newAnswerGroups);
+      sortQuestions: function(sIndex, dIndex) {
+        var stateNamesInOrder = data.questionList.getAllStateNames();
+        var sStateName = stateNamesInOrder[sIndex];
+        var sState = SimpleEditorShimService.getState(sStateName);
+        var dStateName = stateNamesInOrder[dIndex];
+        var dState = SimpleEditorShimService.getState(dStateName);
+        if (dState.interaction.answerGroups.length === 0 ||
```

Comment:
This looks like a long, complicated method that's hard to maintain and test (and in fact I don't see any tests for the new functionality). Assuming that the aim of the method is "move a question from oldIndex to newIndex", how about breaking it up into the following stages (each with their own methods)?

- Define a new tmpQuestion whose properties are identical to the question at oldIndex.
- Delete the question at position oldIndex.
- Insert tmpQuestion into newIndex (or newIndex - 1, as appropriate).

This gives you the added benefit of having an insertQuestion(index) method, which is likely to be useful later if it doesn't already exist. It also makes the flow clearer.

Reply:
Not sure, about this approach. Although I have cleaned up the code according to above comments, following this sort of approach will mean removing other code. Also, I'm not sure about insert questions, basically, we would need to delete two question and insert them twice. 
This is the approach I'm currently thinking of. Let me know if this can be a good way of implementing it.
1.) - Store temp question.
2.) - change answer group destination.
3.) - change incoming nodes from answer group dest.
```        var stateNamesInOrder = data.questionList.getAllStateNames();
        var oldQuestion = SimpleEditorShimService.getState(stateNamesInOrder[oldIndex]);
        var newQuestion = SimpleEditorShimService.getState(stateNamesInOrder[newIndex]);
        if (oldIndex > newIndex){
          // Swap Index to maintain consistency. As question with lower index will have answer group.
          var tempIndex = oldIndex;
          oldIndex = newIndex;
          newIndex = tempIndex;
        }
        var newQuestionAnswerGroups = newQuestion.interaction.answerGroups;
        var oldQuestionAnswerGroups = oldQuestion.interaction.answerGroups;
        oldQuestionAnswerGroups[0].outcome.dest = newQuestionAnswerGroups[0].outcome.dest;
        SimpleEditorShimService.saveAnswerGroups(stateNamesInOrder[oldIndex], oldQuestionAnswerGroups);

        newQuestionAnswerGroups[0].outcome.dest = tempAnswerGroups[0].outcome.dest;
        SimpleEditorShimService.saveAnswerGroups(stateNamesInOrder[newIndex], newQuestionAnswerGroups);
```
**3rd point is still not implemented***

Reply:
Hm, please could you be more precise about the steps you mention? For example, when you say "change answer group destination" and "change incoming nodes from answer group dest", I don't know which answer group you're talking about and whether you're doing this on the temp question or the original question or something else. I'm also still not 100% clear on what the precise aim of this method is, so if you could clarify that, it would help me give better advice. Thanks!

Reply:
Also, I looked at the code snippet you included, and I'm a bit confused. It seems to me that you start with:

      1 --> 2 --> OLD_QUESTION --> 4 --> 5 --> NEW_QUESTION --> 7 --> 8

(say), and then you end up with two chains if you do the destination reassignment as you're envisioning, IIUC:

      1 --> 2 --> OLD_QUESTION --> 7 --> 8
      4 --> 5 --> NEW_QUESTION --> 4 (loop)

I'm not sure this is what you want to end up with? That's why it's important for me to understand what exactly the method is supposed to be doing...